File: 32_Diamond/java/Diamond.java
import java.util.Scanner;

/**
 * Game of Diamond
 * <p>
 * Based on the BASIC game of Diamond here
 * https://github.com/coding-horror/basic-computer-games/blob/main/32%20Diamond/diamond.bas
 * <p>
 * Note:  The idea was to create a version of the 1970's BASIC game in Java, without introducing
 * new features - no additional text, error checking, etc has been added.
 * 
 * Converted from BASIC to Java by Darren Cardenas.
 */
 
public class Diamond {    
  
  private static final int LINE_WIDTH = 60;
  
  private static final String PREFIX = "CC";
  
  private static final char SYMBOL = '!';
  
  private final Scanner scan;  // For user input  
  
  
  public Diamond() {
    
    scan = new Scanner(System.in);    
    
  }  // End of constructor Diamond 
  
  
  public void play() {

    showIntro();
    startGame();

  }  // End of method play  
  
  
  private void showIntro() {    
    
    System.out.println(" ".repeat(32) + "DIAMOND");
    System.out.println(" ".repeat(14) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
    System.out.println("\n\n");
    
  }  // End of method showIntro      


  private void startGame() {        
      
    int body = 0;
    int column = 0;
    int end = 0;
    int fill = 0;
    int increment = 2;    
    int numPerSide = 0;    
    int prefixIndex = 0;
    int row = 0;       
    int start = 1;    
    int userNum = 0;        
    
    String lineContent = "";
    
    // Get user input
    System.out.println("FOR A PRETTY DIAMOND PATTERN,");
    System.out.print("TYPE IN AN ODD NUMBER BETWEEN 5 AND 21? ");
    userNum = scan.nextInt();
    System.out.println("");
    
    // Calcuate number of diamonds to be drawn on each side of screen
    numPerSide = (int) (LINE_WIDTH / userNum);    

    end = userNum;    
    
    // Begin loop through each row of diamonds
    for (row = 1; row <= numPerSide; row++) {
     
      // Begin loop through top and bottom halves of each diamond
      for (body = start; increment < 0 ? body >= end : body <= end; body += increment) {

        lineContent = "";
        
        // Add whitespace
        while (lineContent.length() < ((userNum - body) / 2)) {
          lineContent += " ";
        }
        
        // Begin loop through each column of diamonds
        for (column = 1; column <= numPerSide; column++) {
          
          prefixIndex = 1;
          
          // Begin loop that fills each diamond with characters
          for (fill = 1; fill <= body; fill++) {
            
            // Right side of diamond
            if (prefixIndex > PREFIX.length()) {
              
              lineContent += SYMBOL; 
              
            }
            // Left side of diamond
            else {
              
              lineContent += PREFIX.charAt(prefixIndex - 1);
              prefixIndex++;
              
            }           
            
          }  // End loop that fills each diamond with characters
          
          // Column finished
          if (column == numPerSide) {
            
            break;
            
          }
          // Column not finishd
          else {
            
            // Add whitespace
            while (lineContent.length() < (userNum * column + (userNum - body) / 2)) {
              lineContent += " ";
            }
            
          }          
          
        }  // End loop through each column of diamonds
        
        System.out.println(lineContent);

      }  // End loop through top and bottom half of each diamond

      if (start != 1) {

        start = 1;
        end = userNum;
        increment = 2;        
      
      }
      else {
          
        start = userNum - 2;
        end = 1;
        increment = -2;        
        row--;

      }        

    }  // End loop through each row of diamonds
    
  }  // End of method startGame
  
  
  public static void main(String[] args) {
    
    Diamond diamond = new Diamond();
    diamond.play();
    
  }  // End of method main

}  // End of class Diamond

File: 32_Diamond/kotlin/diamond.kt

/**
 * Game of Diamond
 * <p>
 * Based on the BASIC game of Diamond
 * https://github.com/coding-horror/basic-computer-games/blob/main/32%20Diamond/diamond.bas
 * <p>
 *
 * Changes From Original: Input is validated.
 *
 * Converted from BASIC to Kotlin by Martin Marconcini (@Gryzor)
 * Inspired in the Java code written by Darren Cardenas.
 */


fun main() {
	Diamond().startGame()
}

class Diamond {
	init {
		printIntro()
	}

	fun startGame() {
		var body: Int
		var end: Int
		var start: Int = 1
		var row: Int = 1
		var numPerRow: Int
		var increment: Int = 2
		var lineContent: String
		var prefixIndex: Int
		
		printPrompt()

		// Read the user input
		val input = readLine()

		// Validate input
		val userInput: Int = try {
			input?.toInt() ?: -1
		} catch (e: NumberFormatException) {
			-1
		}
		if (!isValid(userInput)) {
			printInvalidInput()
			return
		}

		// Calculate how many diamonds can horizontally fit in the given space
		numPerRow = calculateDiamondsPerRow(userInput)
		end = userInput

		// Loop throw rows of Diamonds
		while (row <= numPerRow) {
			body = start
			while (canLoop(increment, body, end)) {
				lineContent = ""
				
				// Add white spaces to the "left" of the leftmost diamond.
				while (lineContent.length < ((userInput - body) / 2)) {
					lineContent += " "
				}

				// Begin loop through each column of diamonds
				for (col in 1..numPerRow) {
					prefixIndex = 1

					// Begin loop that fills each diamond with characters (not whitespace)
					for (fill in 1..body) {
						// Right side of diamond, if the index is greater than the prefix, put a Symbol.
						if (prefixIndex > PREFIX.length) {
							lineContent += SYMBOL
						}
						// Left side of diamond, pick a Prefix character (-1 since it starts at 0)
						else {
							lineContent += PREFIX[prefixIndex - 1]
							prefixIndex++
						}
					}// End loop that fills each diamond with characters

					// Is Column finished?
					if (col == numPerRow) {
						break
					}
					// Column is not finished...
					else {
						// Add whitespace on the "right" side of the current diamond, and fill the "left" side of the
						// next; doesn't fill the space to the right of the rightmost diamond.
						while (lineContent.length < (userInput * col + (userInput - body) / 2)) {
							lineContent += " "
						}

					}

				}// End loop through each column of diamonds

				// Print the current Line
				println(lineContent)

				// Increment the body that moves
				body += increment
			} //end While Loop throw rows of Diamonds

			// Increment the current Row of diamonds.
			row++

			if (start != 1) {
				start = 1
				end = userInput
				increment = 2
			} else {
				// We're rendering the "bottom half" of the total rendering.
				// Alter the parameters, and decrease the row number so the logic can loop again.
				start = userInput - 2
				end = 1
				increment = -2
				row--
			}
		} // End loop through each row of diamonds
	}

	private fun canLoop(increment: Int, body: Int, end: Int): Boolean = if (increment < 0) body >= end else body <= end

	private fun calculateDiamondsPerRow(totalDiamonds: Int): Int = LINE_WIDTH / totalDiamonds

	private fun isValid(input: Int): Boolean = (input in 5..21) && (input % 2 != 0)

	private fun printInvalidInput() = println("Invalid input")

	private fun printPrompt() {
		println(
			"""
			FOR A PRETTY DIAMOND PATTERN
			TYPE IN AN ODD NUMBER BETWEEN 5 AND 21? 
		""".trimIndent()
		)
		println()
	}

	private fun printIntro() {
		println(" ".repeat(32) + "DIAMOND")
		println(" ".repeat(14) + "CREATIVE COMPUTING MORRISTOWN, NEW JERSEY")
		println()
		println()
	}

	companion object {
		const val LINE_WIDTH = 60
		const val PREFIX = "CC"
		const val SYMBOL = "!"
	}
}
